New post: Intel and Hugging Face Partner to Democratize Machine Learning Hardware Acceleration (#366)
huggingface/blog: intel.md

---
title: "Intel and Hugging Face Partner to Democratize Machine Learning Hardware Acceleration"
thumbnail: /blog/assets/XXX
---

<h1>Intel and Hugging Face Partner to Democratize Machine Learning Hardware Acceleration</h1>

<div class="blog-metadata">

>     <small>Published May 16, 2022.</small>

Reviewer: Needs an update here :thinking:

With Optimum Intel, you can apply state-of-the-art optimization techniques to your Transformers with minimal effort. Let’s look at a complete example.

## Case study: Quantizing DistilBERT with Optimum Intel

In this example, we will run post-training quantization on a DistilBERT model fine-tuned for classification. Quantization is a process that shrinks memory and compute requirements by reducing the bit width of model parameters. For example, you can often replace 32-bit floating-point parameters with 8-bit integers at the expense of a small drop in prediction accuracy.

We have already fine-tuned the original model to classify product reviews for shoes according to their star rating (from 1 to 5 stars). You can view this [model](https://huggingface.co/juliensimon/distilbert-amazon-shoe-reviews) and its [quantized](https://huggingface.co/juliensimon/distilbert-amazon-shoe-reviews-quantized?) version on the Hugging Face hub. You can also test the original model in this [Space](https://huggingface.co/spaces/juliensimon/amazon-shoe-reviews-spaces). 

> Let’s get started! All code is available in this [notebook](https://gitlab.com/juliensimon/huggingface-demos/-/blob/main/amazon-shoes/03_optimize_inc_quantize.ipynb). 

Reviewer 2: The notebook code is for optimum-intel version < 1.2.3, can I help update it ?

Author: updated.

pip -q install torch==1.11.0+cpu --extra-index-url https://download.pytorch.org/whl/cpu
pip -q install transformers datasets optimum[intel] evaluate --upgrade
```

Then, we prepare an evaluation dataset to assess model performance during quantization. Starting from the dataset we used to fine-tune the original model, we only keep a few thousand reviews and their labels and save them to local storage.

Next, we load the original model, its tokenizer, and the evaluation dataset from the Hugging Face hub.

```
model_name = "juliensimon/distilbert-amazon-shoe-reviews" 
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=5)
tokenizer = AutoTokenizer.from_pretrained(model_name) 
eval_dataset = load_from_disk("./data/amazon_shoe_reviews_test")
```

Next, we define an evaluation function that computes model metrics on the evaluation dataset. This allows the Optimum Intel library to compare these metrics before and after quantization. For this purpose, the Hugging Face [evaluate](https://github.com/huggingface/evaluate/) library is very convenient!

```
from evaluate import evaluator

def eval_func(model): 
    eval = evaluator("text-classification") 

>     results = eval.compute(
>         model_or_pipeline=model, tokenizer=tokenizer, 
>         data=eval_dataset, metric=evaluate.load("accuracy"), 
>     label_column="labels", label_mapping=model.config.label2id
>     ) 

Reviewer 2: ```suggestion
    results = eval.compute(
        model_or_pipeline=model,
        tokenizer=tokenizer,
        data=eval_dataset,
        metric=evaluate.load("accuracy"),
        label_column="labels",
        label_mapping=model.config.label2id,
    )
```

    return results["accuracy"]
```

We then set up the quantization job using a [configuration file](https://huggingface.co/juliensimon/distilbert-amazon-shoe-reviews/blob/main/quantize.yml) that we download from the Hugging Face hub. You can find details on this file in the INC [documentation](https://intel.github.io/neural-compressor/docs/tuning_strategies.html). Here, we go for post-training dynamic quantization with an acceptable accuracy drop of 3%.

```
quantization:
    approach: post_training_dynamic_quant
tuning:
    accuracy_criterion:
        relative: 0.03               
```

Next, we create the corresponding quantization objects with the Optimum Intel API.

```
from optimum.intel.neural_compressor import IncOptimizer, IncQuantizer
from optimum.intel.neural_compressor.configuration import IncQuantizationConfig

quantization_config = IncQuantizationConfig.from_pretrained(
    config_name_or_path="juliensimon/distilbert-amazon-shoe-reviews",
    config_file_name="quantize.yml"
)

> inc_quantizer = IncQuantizer(model, quantization_config, eval_func=eval_func)

Reviewer 2: ```suggestion
inc_quantizer = IncQuantizer(quantization_config, eval_func=eval_func)
```

```

We can now launch the quantization job.

```
inc_optimizer = IncOptimizer(model, quantizer=inc_quantizer)
inc_model = inc_optimizer.fit()
```

> The log tells us that Optimum Intel has quantized 39 ```Linear``` operators.

Reviewer 2: ```suggestion
The log tells us that Optimum Intel has quantized 38 ```Linear``` operators.
```

)
```

```
# Quantized model

TransformerBlock(
  (attention): MultiHeadSelfAttention(
    (dropout): Dropout(p=0.1, inplace=False)
    (q_lin): DynamicQuantizedLinear(in_features=768, out_features=768, dtype=torch.qint8, qscheme=torch.per_channel_affine)
    (k_lin): DynamicQuantizedLinear(in_features=768, out_features=768, dtype=torch.qint8, qscheme=torch.per_channel_affine)
    (v_lin): DynamicQuantizedLinear(in_features=768, out_features=768, dtype=torch.qint8, qscheme=torch.per_channel_affine)
    (out_lin): DynamicQuantizedLinear(in_features=768, out_features=768, dtype=torch.qint8, qscheme=torch.per_channel_affine)
  )
  (sa_layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
  (ffn): FFN(
    (dropout): Dropout(p=0.1, inplace=False)
    (lin1): DynamicQuantizedLinear(in_features=768, out_features=3072, dtype=torch.qint8, qscheme=torch.per_channel_affine)
    (lin2): DynamicQuantizedLinear(in_features=3072, out_features=768, dtype=torch.qint8, qscheme=torch.per_channel_affine)
  )
  (output_layer_norm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
)
```

Very well, but how does this impact accuracy and prediction time?

> Before and after each quantization step, Optimum Intel runs the evaluation function on the current model. Interestingly, the accuracy of the quantized model is now a bit higher (```0.6906```) than the original model (```0.6893```). Likewise, we see that the quantized model predicts the evaluation set 13% faster than the original model. Not bad for a few lines of code!

Author: Interesting, I didn't know that! I'll add it to the post.

```
[INFO] |***********************Tune Result Statistics**********************|
[INFO] +--------------------+-----------+---------------+------------------+
[INFO] |     Info Type      |  Baseline | Tune 1 result | Best tune result |
[INFO] +--------------------+-----------+---------------+------------------+
[INFO] |      Accuracy      |  0.6893   |    0.6906     |     0.6906       |
[INFO] | Duration (seconds) | 106.3149  |    92.4634    |     92.4634      |
[INFO] +--------------------+-----------+---------------+------------------+
```

Accuracy didn’t drop, and Optimum Intel stopped the quantization job after the first step. Had accuracy dropped more than the allowed 3%, Optimum Intel would have tried to quantize different parts of the models until it would have reached an acceptable drop, or the maximum number of trials.

Finally, we save the model and its configuration file to local storage.

```
model_dir = "./model_inc"

> inc_model.model.save_pretrained(model_dir)

Reviewer 2: ```suggestion
inc_optimizer.save_pretrained(model_dir)
```

```

Once we’ve created a new [repository](https://huggingface.co/juliensimon/distilbert-amazon-shoe-reviews-quantized) on the Hugging Face hub and pushed the quantized model to it, we can load it in the usual way.

```
from optimum.intel.neural_compressor.quantization import IncQuantizedModelForSequenceClassification

> inc_model = IncQuantizedModelForSequenceClassification.from_pretrained(
>     "juliensimon/distilbert-amazon-shoe-reviews-quantized")

Reviewer 2: ```suggestion
inc_model = IncQuantizedModelForSequenceClassification.from_pretrained(
    "juliensimon/distilbert-amazon-shoe-reviews-quantized"
)
```
